### ISSUE TITLE
ConnectionDetails for external providers in Spring Security modules

### ISSUE THREAD

**Issue by ParkerM** [spring-projects/spring-boot#36777]
Part of my dev workflow involves using Testcontainers with Keycloak for integration testing against a local, preconfigured IdP/authorization server. It would be awesome if I could use the very same configuration with a running instance through spring-boot-testcontainers, but from what I can tell this isn't currently possible unless the respective configurators implement or expose a `ConnectionDetails` in some way. (I assume I could achieve something like it if I replace the Spring Security autoconfigurations with my own, but I'd prefer to stick with conventions as much as possible.)

The use cases that come to mind are:
 - OAuth2 client provider(s) - URLs defined under `spring.security.oauth2.client.provider`
 - SAML asserting party - Metadata URI defined under `spring.security.saml2.relyingparty.registration.<id>.assertingparty`
 - Probably a few other properties/modules I'm missing

Does this seem like a sensible use case for the ConnectionDetails abstraction? I've noticed the current implementations seem concerned with strictly persistent connections, so perhaps this does not fit the intent. But, since the provider autoconfigurations can lead to startup failure when the server is unreachable, the connections at least have some sort of persistent essence.

**Comment by philwebb** [comment]
@jgrandja @jzheaux Do you have any thoughts on this?

**Comment by philwebb** [comment]
@ParkerM Can you share a bit more information about how you use Testcontainers with Keycloak? Are you using https://testcontainers.com/modules/keycloak/ ?

**Comment by mhalbritter** [comment]
We talked about this yesterday and we think the `ConnectionDetails` abstraction fits for something like this, too.

**Comment by ParkerM** [comment]
> I've not looked at the Keycloak testcontainer in detail, but I remember from the standalone Keycloak that one has to do some realm / client setup before it's usable.

Yeah, I currently have realm import do most of the work then use the Keycloak client to adjust some ports before the tests run (maybe the future will see something like a configurer DSL that simplifies test setup). 

My original plan was to just implement my own `ContainerConnectionDetails` and such, which led me to realize `ConnectionDetails` would be the blocker unless I forego Spring Security AutoConfigurations.

**Comment by Kehrlann** [comment]
Would definitely like to see this happen as well, for any generic OpenID or OAuth2 provider. I'm specifically thinking about [Dex](https://dexidp.io/) and custom-built authorization servers.

I'd want `client_id` and `client_secret` (optionally) available on the `ConnectionDetails`.

**Comment by jgrandja** [comment]
I'm just seeing this now @philwebb. Apologies for the delay.

@rwinch and I have been talking about introducing integration testing support ([spring-authorization-server#258](https://github.com/spring-projects/spring-authorization-server/issues/258)) for Spring Authorization Server.

I'm not familiar with the ConnectionDetails abstraction so I would need to look into this and see if this is an entry point for the support.

**Comment by wilkinsona** [comment]
Moving into 3.4.x to consider alongside #39391 which doesn't tackle Testcontainers or Spring Authorisation Server.

**Comment by wilkinsona** [comment]
I've been looking at https://github.com/spring-projects/spring-boot/pull/39391 which tackles part of this, and I'm struggling to figure out exactly how broad the connection details should be. It's not clear to me where we should draw the line in terms of what the connection details should override and what should come from the normal configuration.

Just focussing on the OAuth2 client side of things for now, @ParkerM, @Kehrlann, @PhilKes, could you please provide some more details about how you're configuring things at the moment (the application.properties or application.yaml) with `spring.security.oauth2.client.registration.*` and `spring.security.oauth2.provider.*` properties and then what you'd typically do with Keycloak to test things that are set up using these properties. I'm also curious if you'd only ever use Keycloak managed by Docker Compose or Testcontainers when the app would also use Keycloak in production, or if you try to use Keycloak as an easy-to-run-in-a-container substitute for something else.

**Comment by ParkerM** [comment]
> could you please provide some more details about how you're configuring things at the moment (the application.properties or application.yaml) with `spring.security.oauth2.client.registration.*` and `spring.security.oauth2.provider.*` properties and then what you'd typically do with Keycloak to test things that are set up using these properties.

In my case I was working on a federated login sort of thing that used spring authorization server as a primary OAuth2 provider and Keycloak as an "external" IdP. So, while I did need to set some `registration` properties at runtime to configure the complicated test, I think that's more specific to the peculiar nature of the project. 

I didn't really foresee `registration` properties being in scope for `ConnectionDetails` because they either require configuration at the IdP (e.g., establishing `redirect_uri` for auth code grant), or are options that the app itself selects (e.g., requested scopes, grant type, client authentication method).

On the other hand, `provider` properties seem like a good fit because they are merely properties of the IdP itself, rather than properties of an established relationship between the application and the IdP. `ConnectionDetails` would be a convenient critical path from which IdP properties are consumed, be it physical properties like URL's or declared properties like supported algorithms — basically anything that the application must deal with or die.

In my particular case I was using Keycloak as a SAML provider, so I was interested in configuring the `assertingparty.metadata-uri` property. There are also many other provider-esque properties under the saml2 `assertingparty` namespace that can be used in lieu of a `metadata-uri`, but they lack the clear boundaries we get with oauth2's `registration` and `provider` so I'll skip them to avoid being distracting.

> I'm also curious if you'd only ever use Keycloak managed by Docker Compose or Testcontainers when the app would also use Keycloak in production, or if you try to use Keycloak as an easy-to-run-in-a-container substitute for something else.

We support Keycloak as a provider and occasionally use it in production-like environments, but in this case I selected it for easy-to-run-in-a-container reasons. Namely for the convenience of realm export/import as a way to "serialize" a production instance so it can be reproduced in an ephemeral container with little fuss and minor additional configuration. The additional configuration not handled by the realm file boils down to (re-)configuring callback URI's via the Keycloak API.
